source/slang/slang-lower-to-ir.cpp

```diff
@@ -6228,6 +6310,79 @@ struct DeclLoweringVisitor : DeclVisitor<DeclLoweringVisitor, LoweredValInfo>
         return Slang::maybeGetConstExprType(getBuilder(), type, decl);
     }
 
+        /// Emit appropriate generic parameters for a constraint, and return the value of that constraint.
+        ///
+        /// The `supType` paramete represents the super-type that a parameter is constrained to.
+    IRInst* emitGenericConstraintValue(
+        IRGenContext*               subContext,
+        GenericTypeConstraintDecl*  constraintDecl,
+        IRType*                     supType)
+    {
+        auto subBuilder = subContext->irBuilder;
+
+        // There are two cases we care about here.
+        //
+        if(auto andType = as<IRConjunctionType>(supType))
+        {
+            // The non-trivial case is when the constraint on a generic parameter
+            // was of the form `T : A & B`. In this case, we really want to
+            // emit the function with parameters for each of the two independent
+            // constraints `T : A` and `T : B`.
+            //
+            // We will loop over the "cases" of the conjunction (since
+            // the `IRConunctionType` can support more than just binary
+            // conjunctions) and recursively add constraints for each.
+            //
+            List<IRInst*> caseVals;
+            auto caseCount = andType->getCaseCount();
+            for(Int i = 0; i < caseCount; ++i)
+            {
+                auto caseType = andType->getCaseType(i);
+                auto caseVal = emitGenericConstraintValue(subContext, constraintDecl, caseType);
+                caseVals.add(caseVal);
+            }
+
+            return subBuilder->emitMakeTuple(caseVals);
+        }
+        else
+        {
+            // The case case is any other type being used as the constraint.
+            //
+            // The constraint will then map to a single generic parameter passing
+            // a witness table for conformance to the given `supType`.
+            //
+            auto param = subBuilder->emitParam(subBuilder->getWitnessTableType(supType));
+            addNameHint(context, param, constraintDecl);
+
+            // In order to support some of the "any-value" work in dynamic dispatch
+            // we have to attach the interface that was used as a constraint onto the
+            // type that is being constrained (which we expect to be a generic type
+            // parameter).
+            //
+            // TODO: It feels a bit gross to be doing this here; perhaps the front-end
```

Comment:
This definitely seems something better done in the front-end. And we probably don't need the general `TypeConstraintDecoration`, and just use a simple `ValueSizeDecoration`.